### PR TITLE
Round robin during spread scheduling

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?

- Separate spread scheduling and default hydra scheduling (i.e. SpreadScheduling != HybridScheduling(threshold=0)): they are already separated in the API layer and they have the different end goals so it makes sense to separate their implementations and evolve them independently.
- Simple round robin for spread scheduling: this is just a starting implementation, can be optimized later.
- Prefer not to spill back tasks that are waiting for args since the pull is already in progress.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by scv119 — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by jjyao** [comment]
pipelined_ingestion_1500_gb_15_windows  passed: https://buildkite.com/ray-project/periodic-ci/builds/2814

**Comment by jjyao** [comment]
Core nightly looks good: https://buildkite.com/ray-project/periodic-ci/builds/2821

**Comment by scv119** [comment]
looks very simple and clean. @ericl do you want to take another look?

**Review by scv119 — APPROVED** [review]

File: python/ray/data/impl/shuffle.py [hunk 1]
```diff
@@ -31,6 +31,8 @@ def simple_shuffle(
         map_ray_remote_args = {}
     if reduce_ray_remote_args is None:
         reduce_ray_remote_args = {}
+    if "scheduling_strategy" not in reduce_ray_remote_args:
+        reduce_ray_remote_args["scheduling_strategy"] = "SPREAD"
```
[the remark this answers is not in the document]
Comment (jjyao) [reply python/ray/data/impl/shuffle.py:35]: Yea, you can explicitly set `scheduling_strategy="DEFAULT"` in this case.

**Pull request merged**